### DISCUSSION TITLE
Keep-alive issue ?

### DISCUSSION THREAD

**Discussion by drannou** [jmcollin78/versatile_thermostat#379] [Q&A] [answered]
Hello
I'm facing regulary the same issue with Versatile, may be someone knows what's going on
I have sometime versatile that is stuck in radiator "ON" mode, even if the temperature is upper than the set value.

Even if I switch off the thermostat, and manually stop the heater, it's forcing back on:

```
2024-02-13 13:47:57.261 DEBUG (MainThread) [custom_components.versatile_thermostat.keep_alive] Calling keep-alive action
2024-02-13 13:47:57.261 DEBUG (MainThread) [custom_components.versatile_thermostat.underlyings] VersatileThermostat-thermostat_lucie-switch.heater_XXX - Starting underlying entity switch.heater_XXXX
```
The only possible for me is to disable (in HA) this thermostat, force it off, and re enable

Does anyone knows how to fix that (beffore putting. a lot of debug in the code :p )

**Comment by jmcollin78** [comment]
Hello @drannou ,

This is not normal. Can you please post the attributes of your VTherm because there is many types, modes, ... and I can't do anything without seeing your configuration. Is it available in the Development Tools / States. You should have something like this:

https://github.com/jmcollin78/versatile_thermostat/blob/main/README.md#custom-attributes

More over, you can have a look here if you find your case : https://github.com/jmcollin78/versatile_thermostat/blob/main/README.md#the-radiator-heats-up-even-though-the-setpoint-temperature-is-exceeded-or-does-not-heat-up-even-though-the-room-temperature-is-well-below-the-setpoint

**Reply by drannou** [reply]
I'm in a standard things, with a preaty well isolated. It's working like a charm most of the time, except when I have this type of bug. Please also note that I have 2 rooms working like that, and have the same bug (not at the same time :) )

For the heater action, I'm using a switch controling the pilot mode of the heater

```
 switches:

    heater_XXX:
        friendly_name: heater_XXX
        value_template: "{{ is_state('switch.heater_XXX', 'off') }}"
        turn_on:
          service: select.select_option
          data:
            option: 'comfort'
          target:
            entity_id: select.heat_XXX_pilot_wire_mode
        turn_off:
          service: select.select_option
          data:
            option: 'off'
          target:
            entity_id: select.heat_XXX_pilot_wire_mode
```

Currently debug values (no bug for the moment):
```
hvac_modes: heat, off
min_temp: 7
max_temp: 25
target_temp_step: 0.1
preset_modes: none, frost, eco, comfort, boost
current_temperature: 19.5
temperature: 18
hvac_action: off
preset_mode: eco
is_on: false
hvac_mode: off
type: null
is_controlled_by_central_mode: true
last_central_mode: null
frost_temp: 10
eco_temp: 18
boost_temp: 20
comfort_temp: 19
frost_away_temp: 7
eco_away_temp: 7
boost_away_temp: 7
comfort_away_temp: 7
power_temp: 13
target_temperature_step: 0.1
ext_current_temperature: 12.4
ac_mode: false
current_power: null
current_power_max: null
saved_preset_mode: eco
saved_target_temp: 18.4
saved_hvac_mode: null
motion_sensor_entity_id: null
motion_state: null
power_sensor_entity_id: null
max_power_sensor_entity_id: null
overpowering_state: null
presence_sensor_entity_id: null
presence_state: null
window_state: off
window_auto_state: off
window_bypass_state: false
window_sensor_entity_id: binary_sensor.window_XXX_contact
window_delay_sec: 30
window_auto_enabled: false
window_auto_open_threshold: null
window_auto_close_threshold: null
window_auto_max_duration: null
window_action: window_turn_off
security_delay_min: 60
security_min_on_percent: 0.5
security_default_on_percent: 0.1
last_temperature_datetime: 2024-02-13T21:37:12.390402+01:00
last_ext_temperature_datetime: 2024-02-13T21:13:45.443738+01:00
security_state: false
minimal_activation_delay_sec: 10
device_power: 1200
mean_cycle_power: 0
total_energy: 26352
last_update_datetime: 2024-02-13T21:37:12.392839+01:00
timezone: CET
temperature_unit: °C
is_device_active: false
ema_temp: 19.68
is_used_by_central_boiler: false
is_over_switch: true
is_inversed: false
underlying_switch_0: switch.heater_XXX
underlying_switch_1: null
underlying_switch_2: null
underlying_switch_3: null
on_percent: 0
power_percent: 0
on_time_sec: 0
off_time_sec: 300
cycle_min: 5
function: tpi
tpi_coef_int: 0.6
tpi_coef_ext: 0.01
friendly_name: thermostat_XXX
supported_features: 17
```

**Comment by jmcollin78** [comment]
Have you configured the keep-alive ? If you don't need it, you should not use it.

@pdcastro can you have a look please and try to understand why there is this log:
```[custom_components.versatile_thermostat.keep_alive] Calling keep-alive action``` whenever the VTherm seems not to have the keep-alive enabled ? (I don't see the value in attributes).

@drannou , I need the attributes, when the bug is currently running. Somethinh seems to force the heater to be on, and I need to understand what does this.

**Reply by pdcastro** [reply]
@jmcollin78, `keep_alive_sec` was not showing among the VTherm climate attributes in the Home Assistant Developer Tools > States area because it had not been exposed in the code. I have just submitted PR #381 that exposes the attribute and also improves the log message a little bit. With the new code, we will have a bit more information to go by. I currently assume (but cannot say for sure) that the user enabled the keep-alive feature by entering a positive number of seconds in the config flow form.

**Comment by drannou** [comment]
I was able to reproduce by forcing in versatile a high temperature (22°C) while temperature was low (18), even stopping versatile, it's putting back the heater in comfort mode

```
2024-02-14 10:53:26.828 INFO (MainThread) [custom_components.versatile_thermostat.base_thermostat] VersatileThermostat-thermostat_lucie - Set hvac mode: off
2024-02-14 10:53:26.828 DEBUG (MainThread) [custom_components.versatile_thermostat.base_thermostat] VersatileThermostat-thermostat_lucie - Checking new cycle. hvac_mode=off, security_state=False, preset_mode=eco
2024-02-14 10:53:26.828 DEBUG (MainThread) [custom_components.versatile_thermostat.base_thermostat] VersatileThermostat-thermostat_lucie - Window auto is on, check the alert. last slope is 1.050
2024-02-14 10:53:26.829 DEBUG (MainThread) [custom_components.versatile_thermostat.base_thermostat] VersatileThermostat-thermostat_lucie - Window auto event is ignored because bypass is ON or window auto detection is disabled
2024-02-14 10:53:26.829 DEBUG (MainThread) [custom_components.versatile_thermostat.base_thermostat] VersatileThermostat-thermostat_lucie - power not configured. check_overpowering not available
2024-02-14 10:53:26.829 DEBUG (MainThread) [custom_components.versatile_thermostat.base_thermostat] VersatileThermostat-thermostat_lucie - checking security delta_temp=4.2 delta_ext_temp=13.6 mod_cond=False temp_cond=False climate_cond=False switch_cond=False
2024-02-14 10:53:26.829 DEBUG (MainThread) [custom_components.versatile_thermostat.base_thermostat] VersatileThermostat-thermostat_lucie - End of cycle (HVAC_MODE_OFF)
2024-02-14 10:53:26.829 DEBUG (MainThread) [custom_components.versatile_thermostat.base_thermostat] VersatileThermostat-thermostat_lucie - last_change_time is now 2024-02-14 11:53:26.829616+01:00
2024-02-14 10:53:26.831 DEBUG (MainThread) [custom_components.versatile_thermostat.thermostat_switch] VersatileThermostat-thermostat_lucie - Calling update_custom_attributes: {'is_on': False, 'hvac_action': <HVACAction.OFF: 'off'>, 'hvac_mode': <HVACMode.OFF: 'off'>, 'preset_mode': 'eco', 'type': None, 'is_controlled_by_central_mode': True, 'last_central_mode': None, 'frost_temp': 10.0, 'eco_temp': 18.0, 'boost_temp': 20.0, 'comfort_temp': 19.0, 'frost_away_temp': 7.0, 'eco_away_temp': 7.0, 'boost_away_temp': 7.0, 'comfort_away_temp': 7.0, 'power_temp': 13.0, 'target_temperature_step': 0.1, 'ext_current_temperature': 12.4, 'ac_mode': False, 'current_power': None, 'current_power_max': None, 'saved_preset_mode': 'eco', 'saved_target_temp': 14.6, 'saved_hvac_mode': None, 'motion_sensor_entity_id': None, 'motion_state': None, 'power_sensor_entity_id': None, 'max_power_sensor_entity_id': None, 'overpowering_state': None, 'presence_sensor_entity_id': None, 'presence_state': None, 'window_state': 'off', 'window_auto_state': 'off', 'window_bypass_state': False, 'window_sensor_entity_id': 'binary_sensor.window_lucie_contact', 'window_delay_sec': 30, 'window_auto_enabled': False, 'window_auto_open_threshold': None, 'window_auto_close_threshold': None, 'window_auto_max_duration': None, 'window_action': 'window_turn_off', 'security_delay_min': 60, 'security_min_on_percent': 0.5, 'security_default_on_percent': 0.1, 'last_temperature_datetime': '2024-02-14T11:53:26.829607+01:00', 'last_ext_temperature_datetime': '2024-02-14T11:53:26.829607+01:00', 'security_state': False, 'minimal_activation_delay_sec': 10, 'device_power': 1200.0, 'mean_cycle_power': 0.0, 'total_energy': 26942.0, 'last_update_datetime': '2024-02-14T11:53:26.830000+01:00', 'timezone': 'CET', 'temperature_unit': <UnitOfTemperature.CELSIUS: '°C'>, 'is_device_active': False, 'ema_temp': 19.14, 'is_used_by_central_boiler': False, 'is_over_switch': True, 'is_inversed': False, 'underlying_switch_0': 'switch.heater_lucie', 'underlying_switch_1': None, 'underlying_switch_2': None, 'underlying_switch_3': None, 'on_percent': 0, 'power_percent': 0, 'on_time_sec': 0, 'off_time_sec': 300, 'cycle_min': 5, 'function': 'tpi', 'tpi_coef_int': 0.6, 'tpi_coef_ext': 0.01}
2024-02-14 10:53:26.831 INFO (MainThread) [custom_components.versatile_thermostat.const] VersatileThermostat-thermostat_lucie - Sending event EventType.HVAC_MODE_EVENT with data: {'hvac_mode': <HVACMode.OFF: 'off'>}
2024-02-14 10:53:26.844 DEBUG (MainThread) [custom_components.versatile_thermostat.sensor] thermostat_lucie_last_temp_datetime - climate state change
2024-02-14 10:53:26.845 DEBUG (MainThread) [custom_components.versatile_thermostat.sensor] thermostat_lucie_last_ext_temp_datetime - climate state change
2024-02-14 10:53:26.845 DEBUG (MainThread) [custom_components.versatile_thermostat.sensor] thermostat_lucie_temperature_slope - climate state change
2024-02-14 10:53:26.846 DEBUG (MainThread) [custom_components.versatile_thermostat.sensor] thermostat_lucie_ema_temperature - climate state change
2024-02-14 10:53:26.846 DEBUG (MainThread) [custom_components.versatile_thermostat.sensor] thermostat_lucie_energy - climate state change
2024-02-14 10:53:26.846 DEBUG (MainThread) [custom_components.versatile_thermostat.sensor] thermostat_lucie_mean_power_cycle - climate state change
2024-02-14 10:53:26.846 DEBUG (MainThread) [custom_components.versatile_thermostat.sensor] thermostat_lucie_power_percent - climate state change
2024-02-14 10:53:26.846 DEBUG (MainThread) [custom_components.versatile_thermostat.sensor] thermostat_lucie_on_time - climate state change
2024-02-14 10:53:26.846 DEBUG (MainThread) [custom_components.versatile_thermostat.sensor] thermostat_lucie_off_time - climate state change
2024-02-14 10:53:26.847 DEBUG (MainThread) [custom_components.versatile_thermostat.binary_sensor] thermostat_lucie_security_state - climate state change
2024-02-14 10:53:26.847 DEBUG (MainThread) [custom_components.versatile_thermostat.binary_sensor] thermostat_lucie_window_bypass_state - climate state change
2024-02-14 10:53:26.847 DEBUG (MainThread) [custom_components.versatile_thermostat.binary_sensor] thermostat_lucie_window_state - climate state change
```

some seconds after:
```
2024-02-14 10:54:28.731 DEBUG (MainThread) [custom_components.versatile_thermostat.underlyings] VersatileThermostat-thermostat_lucie-switch.heater_lucie - Starting underlying entity switch.heater_lucie
2024-02-14 10:56:28.736 DEBUG (MainThread) [custom_components.versatile_thermostat.underlyings] VersatileThermostat-thermostat_lucie-switch.heater_lucie - Starting underlying entity switch.heater_lucie
```

Values:
```
hvac_modes: heat, off
min_temp: 7
max_temp: 25
target_temp_step: 0.1
preset_modes: none, frost, eco, comfort, boost
current_temperature: 19.3
temperature: 18
hvac_action: off
preset_mode: eco
is_on: false
hvac_mode: off
type: null
is_controlled_by_central_mode: true
last_central_mode: null
frost_temp: 10
eco_temp: 18
boost_temp: 20
comfort_temp: 19
frost_away_temp: 7
eco_away_temp: 7
boost_away_temp: 7
comfort_away_temp: 7
power_temp: 13
target_temperature_step: 0.1
ext_current_temperature: 12.4
ac_mode: false
current_power: null
current_power_max: null
saved_preset_mode: eco
saved_target_temp: 14.6
saved_hvac_mode: null
motion_sensor_entity_id: null
motion_state: null
power_sensor_entity_id: null
max_power_sensor_entity_id: null
overpowering_state: null
presence_sensor_entity_id: null
presence_state: null
window_state: off
window_auto_state: off
window_bypass_state: false
window_sensor_entity_id: binary_sensor.window_lucie_contact
window_delay_sec: 30
window_auto_enabled: false
window_auto_open_threshold: null
window_auto_close_threshold: null
window_auto_max_duration: null
window_action: window_turn_off
security_delay_min: 60
security_min_on_percent: 0.5
security_default_on_percent: 0.1
last_temperature_datetime: 2024-02-14T11:53:26.829607+01:00
last_ext_temperature_datetime: 2024-02-14T11:53:26.829607+01:00
security_state: false
minimal_activation_delay_sec: 10
device_power: 1200
mean_cycle_power: 0
total_energy: 26942
last_update_datetime: 2024-02-14T11:53:26.830000+01:00
timezone: CET
temperature_unit: °C
is_device_active: false
ema_temp: 19.14
is_used_by_central_boiler: false
is_over_switch: true
is_inversed: false
underlying_switch_0: switch.heater_lucie
underlying_switch_1: null
underlying_switch_2: null
underlying_switch_3: null
on_percent: 0
power_percent: 0
on_time_sec: 0
off_time_sec: 300
cycle_min: 5
function: tpi
tpi_coef_int: 0.6
tpi_coef_ext: 0.01
friendly_name: thermostat_lucie
supported_features: 17
```

**Reply by jmcollin78** [reply]
For me there no heating requested by VTherm: 
```
on_percent: 0
power_percent: 0
hvac_action: off
hvac_mode: off
``` 

You don't answer my questions: have you enabled the keep-alive feature ?

**Reply by drannou** [reply]
yes Keepalive is up, it's keepalive that is asking for heat
keep-alive is configured fgor a 120s run, which match the logs:
```
2024-02-14 10:54:28.731 DEBUG (MainThread) [custom_components.versatile_thermostat.underlyings] VersatileThermostat-thermostat_lucie-switch.heater_lucie - Starting underlying entity switch.heater_lucie
2024-02-14 10:56:28.736 DEBUG (MainThread) [custom_components.versatile_thermostat.underlyings] VersatileThermostat-thermostat_lucie-switch.heater_lucie - Starting underlying entity switch.heater_lucie

```

**Reply by jmcollin78** [reply]
Why do you activate the keep-alive ? Can you deactivate it ?

@pdcastro I think we have a bug with the keep-alive that turns on a heater while it was off. Can you please have a look ?

**Reply by drannou** [reply]
Good question !
I activate it because I had some times an issue: some times, the heater was stuck in "comfort" mode.
I was not sure if it was from versatile or the zigbee that lost the 'go in off mode' message.
May be the problem is linked, may be not...

**Reply by jmcollin78** [reply]
You should deactivate the keep-alive. This is not done for that purpose but only for device which need this. Deactive and restart , it should works now. And we will see the Zigbee error (another error with no link to keep-alive)

**Reply by pdcastro** [reply]
I have not yet seen hard evidence of a bug in the keep-alive feature. @drannou, I am bit puzzled by the Template Switch definition that you posted [earlier](#discussioncomment-8458599), in particular the `value_template` expression:

```
 switches:
    heater_XXX:
        friendly_name: heater_XXX
        value_template: "{{ is_state('switch.heater_XXX', 'off') }}"
```

Shouldn’t that `'off'` perhaps be `'on'` instead? It sounds like the value of the `heater_XXX` switch is True / On when the state of `'switch.heater_XXX'` is `'off'`. I don’t get it. 😄 The Home Assistant documentation has a few examples, and all examples use `'on'` in that expression:
https://www.home-assistant.io/integrations/switch.template/

**Reply by pdcastro** [reply]
Also, still on the Template Switch definition:
```
        turn_on:
          service: select.select_option
          data:
            option: 'comfort'
          target:
            entity_id: select.heat_XXX_pilot_wire_mode
```
When the template switch is turned on, it selects a ‘comfort’ preset somewhere. You had also previously stated that _“sometimes, the heater was stuck in ‘comfort’ mode”._ 

Is the ‘comfort’ preset being applied to an instance of Versatile Thermostat (VTherm), or is the preset being applied to some other separate [Climate integration](https://www.home-assistant.io/integrations/climate/)?

If it is the former, then I am confused as to why VTherm should apply a preset on itself through the switch `turn_on` service, as opposed to your UI or automation calling the [`set_preset_mode`](https://www.home-assistant.io/integrations/climate/#service-climateset_preset_mode) service.

If it is the latter, then I am confused as to why VTherm is being used in [“thermostat over switch”](https://github.com/jmcollin78/versatile_thermostat/blob/main/README.md#when-to-use--not-use) mode as opposed to “thermostat over climate”. 🤔

**Reply by jmcollin78** [reply]
It is applied on another switch which is linked to the heater via the pilot wire. This is quite common pattern with pilot wire.
The inversion in template above seems more a good clue. Maybe @drannou wants to inverse the command but don't tells us that fact. That is also a common pattern for some pilot wire heater. If yes, you have a native option in VTherm to inverse the command: `is_inversed: false`.

Nevertheless, thank you @pdcastro for your investigations, it gives us some interesting insights. I continue to think that you should not use the keep-alive feature for a workaround about command lost. It was really not implemented for this but for device that automatically turns off when they receive no more command for a certain time. @pdcastro should be able to confirm (or not).

**Reply by drannou** [reply]
Thanks for your remarks, I clean up that part (which was not clear) and remove the keepalive. 
I still think that keep-alive could be a good idea in case of manual intervention, but if you fell it's not a good idea, I disable it :)

Thanks !

**Reply by pdcastro** [reply]
> _”pdcastro should be able to confirm (or not).”_

The keep-alive feature is very simple: It keeps repeating the ***last*** turn on / turn off command issued by the Versatile Thermostat, at regular intervals. Like **jmcollin78** said, the most common use case is that of switches that automatically turn off if they do not receive “turn on” commands for a period of time (failsafe mode). However, I don’t see why it could not be used in other cases. If a user has a wireless switch that occasionally loses a turn on/off command, perhaps because the radio signal is weak, and the command simply needs to be repeated, then it actually sounds to me like a good use case for the keep-alive feature as well.

In an [earlier post](#discussioncomment-8464264), **drannou** shared logs that start with `"Set hvac mode: off"` and end with `"Starting underlying entity"` about a minute later:

```
2024-02-14 10:53:26.828 INFO (MainThread) [custom_components.versatile_thermostat.base_thermostat] VersatileThermostat-thermostat_lucie - Set hvac mode: off
...
2024-02-14 10:54:28.731 DEBUG (MainThread) [custom_components.versatile_thermostat.underlyings] VersatileThermostat-thermostat_lucie-switch.heater_lucie - Starting underlying entity switch.heater_lucie
```

In between the two timestamps above, I don’t see any log messages with`"Calling keep-alive action”`. I also don’t see any log messages with `"Stopping underlying entity"`. @drannou, had you intentionally omitted some such log messages between the two timestamps above?

@jmcollin78, assuming no log messages were omitted, we need to find out why there wasn’t a single “Stopping underlying entity” log message when the VTherm was turned off, and/or what causes the ***first*** “turn on” command to be issued *after* the VTherm is turned off. The keep-alive feature will simply keep repeating the last command, by design.

**Reply by drannou** [reply]
I can confirm you that I did not cut the logs, and I still don't understand why versatil / keepalive was forcing to on while everything say it's off.
I was reading the source code and may be I miss something but I was wondering why there is no test on the `def turn_on(self)` to check if it should really be there (when keepalive is activated)

**Reply by jmcollin78** [reply]
> However, I don’t see why it could not be used in other cases. If a user has a wireless switch that occasionally loses a turn on/off command

I understand but my opinion is that, you should rather fix the root cause of your problem (the weak radio signal or whatever) than using a feature than is not made for this. It seems to me a bad idea to resend all commands multiple times for all equipments because you loose some command. Add Zigbee router if you are in Zigbee, add Wifi repeater, remove Chinese Cloud based eqt, ... and all your network and equipments will work better.
More over, if you cannot join an eqt because it is not reachable, you can sent it command every second, it will stay not reachable.

**Reply by pdcastro** [reply]
> _I was reading the source code and may be I miss something but I was wondering why there is no test on the `def turn_on(self)` to check if it should really be there (when keepalive is activated)_

This has already changed a bit today, through PR #382. In case this explanation helps: There is only one keep-alive callback at any point in time, scheduled through `self._keep_alive.set_async_action(callback)`. It executes in a single concurrency thread through Python’s [asyncio](https://docs.python.org/3/library/asyncio.html) library, in the same IO loop as the Home Assistant core. Before PR #382, the `turn_on(self)` and `turn_off(self)` methods would set the callback to themselves:

* `turn_on(self)` calls `self._keep_alive.set_async_action(self.turn_on)`
* `turn_off(self)` calls `self._keep_alive.set_async_action(self.turn_off)`

Therefore, if the callback calls `turn_on(self)`, it must be because `turn_on(self)` was the last method to have been called (rather than `turn_off(self)`): either called by the callback, or called as a result of the thermostat’s logic (target temperature reached, thermostat turned on/off through a preset – `HVACMode.OFF`...). If `turn_off(self)` was the last method to have been called, then the callback would have been replaced (including the cancellation of any previously scheduled callbacks) to keep calling `turn_off(self)` instead of `turn_on(self)`.

After PR #382, both `turn_on(self)` and `turn_off(self)` schedule the same callback method that includes a test to decide whether to call `turn_on` or `turn_off` at the time when the callback fires, which is probably more intuitive. But I am not convinced that it is a “bug fix” because I have not reproduced or clearly identified a fault with the previous implementation. In a [PR comment](https://github.com/jmcollin78/versatile_thermostat/pull/382/files#r1491868666), I described a hypothetical situation that might be avoided by the new logic, but if it was possible at all, it would be a rare corner case rather than something that happens consistently as you have reported.

**Comment by drannou** [comment]
Please also note taht if I disable/reenable, Versatile is stopping himself the heater

`2024-02-14 11:00:31.970 DEBUG (MainThread) [custom_components.versatile_thermostat.underlyings] VersatileThermostat-thermostat_lucie-switch.heater_lucie - Stopping underlying entity switch.heater_lucie`

**Comment by jmcollin78** [comment]
@drannou , thanks to @pdcastro , there is some improvements for corner case, that has been implemented in the release https://github.com/jmcollin78/versatile_thermostat/releases/tag/5.4.2

Can you please give it a try and let us know if you still have the issue (with keep-alive activated).

**Comment by drannou** [comment]
Hello
So I tested the patch of @pdcastro , it's perfectly working.
I continue to make some tests, and it seems that, for me, when I'm playing with the temperature:
- I set a high temperature, versatile is putting the heater in ON (comfort for me), everything is normal
- Right after, I put a low temperature, versatil is NOT stopping the heater.

I dig a little bit the code and there is something strange for me (sorry I'm a newbie on versatile code):
When I'm putting a low temperature, versatile is re-calculating everything, but display a "nothing to do" which is https://github.com/jmcollin78/versatile_thermostat/blob/main/custom_components/versatile_thermostat/underlyings.py#L343

I'm wondering why it's not going just before in 
https://github.com/jmcollin78/versatile_thermostat/blob/main/custom_components/versatile_thermostat/underlyings.py#L334
I change a little bit the condition, with a
`        elif self.is_device_active or on_time_sec == 0:`
and by doing that, I'm going on the `await self.turn_off()`, and my heater stop

What is exactly the purpose of "def is_device_active" ? I would think that it's there to know if it's currently heating or not, but When I check (and add) some logs, it seems that it's always False

**Reply by pdcastro** [reply, marked as answer]
@drannou, the `is_device_active()` method returns the state of entities like a switch, for example “on” or “off”, _as known by Home Assistant_ (e.g. it calls `self._hass.states.is_state()`), plus some logic to take `self.is_inversed` into account. So yes, like you thought, `is_device_active()` is used to find out whether your heater is currently heating or not. If you find that `is_device_active()` always returns `False`, even when your heater is turned on, that’s a problem that needs to be solved — and it’s a great finding too, halfway to the solution!

I know that some types of switch implementations, such as certain MQTT switch configurations, make a distinction between “turning on/off” and “being on/off”. For example, Home Assistant might send a command (over the network) to the switch to turn on/off, but it does not assume that the switch was actually turned on/off. It then queries the status of the switch — _“has it actually turned on/off as commanded?”_ Calling the `turn_on()` or `turn_off()` methods results in that command being sent to the switch, while the `is_device_active()` method reflects Home Assistants knowledge of whether the switch was actually turned on/off as commanded.

Based on the above, I would propose an experiment: Temporarily remove the VTherm (Versatile Thermostat) integration and use the HASS (Home Assistant) [Developer Tools](https://www.home-assistant.io/docs/tools/dev-tools/) to manually turn the heater switch on/off and observe that its state actually switches on/off. Steps:

* Temporarily remove the Versatile Thermostat integration: Settings ➔ Integrations ➔ Versatile Thermostat ➔ Three vertical dots ➔ Delete.
* Restart Home Assistant: Developer Tools ➔ YAML ➔ Restart
* Manually issue a “turn on” command to the switch: Developer Tools ➔ Services tab ➔ Select the “Switch: Turn on” service ➔ Choose Entity button ➔ Select the switch entity for the heater (the one that isn’t properly turning off by Versatile Thermostat) ➔ Call Service button.
* Observe that the switch state is “on”: Developer Tools ➔ States tab ➔ Type the switch name in the Filter Entities box ➔ Observe the switch state in the State column. The state should be “on”.
* Repeat the previous two steps, now turning the switch “off” and verifying that the state is “off”.

If you find that the switch state is not switching to “off”, then the problem is with the switch configuration, rather than a problem with Versatile Thermostat. If the state becomes “off” just fine, then it might indeed be an issue with Versatile Thermostat that will need to be investigated further.

Regarding that `elif self.is_device_active` logic test, I understand that it’s there just to avoid turning the switch off when it is already off. Some switches might be in cordless, battery-powered devices (e.g. hypothetically a battery-powered Zigbee switch attached to a mains-powered heater) and just processing an unnecessary turn on / turn off command might be a waste of battery plus unnecessary radio noise in shared channels (e.g. 2.4GHz channel shared between WiFi and Zigbee, if I recall correctly). But if somehow we found that we cannot rely on HASS’ state representation for switches, of if you wanted a workaround just for your case, I would say that:

Instead of:
```python
elif self.is_device_active or on_time_sec == 0:
    _LOGGER.info("... stop heating ..." ...)
    await self.turn_off()
else:
    _LOGGER.debug("%s - nothing to do", self)
```
You could just do:
```python
else:
    _LOGGER.info("... stop heating ..." ...)
    await self.turn_off()
```
(No `elif`, no “nothing to do” log statement.)

**Reply by drannou** [reply]
Ok so that was my mistake, I made a wrong configuration in my switch,  the "value_template" was wrongly configured...
Anyway I learn a lot on the algorithm of the plugin!
Thanks for your time !
